File: 0824_goat_latin/to_goat_latin.java
class Solution {
    public String toGoatLatin(String S) {
        Set<Character> vowels = new HashSet<>();
        for (char c : "aeiouAEIOU".toCharArray())
            vowels.add(c);
        String suffix = "";
        StringBuilder sb = new StringBuilder();
        for (String s : S.split(" ")) {
            suffix += 'a'; // Java compiler converts this into a temporary StringBuilder
            if (!vowels.contains(s.charAt(0))) {
                s = s.substring(1) + s.charAt(0);
            }
            sb.append(s).append("ma").append(suffix).append(' ');
        }
        sb.deleteCharAt(sb.length() - 1); // remove extra space
        return sb.toString();
    }
}

File: 0824_goat_latin/to_goat_latin.py
def to_goat_latin(S: str) -> str:
    """
    Time: O(n**2)
    Space: O(n**2)
    """
    split = S.split(" ")
    res = []
    vowels = set(['a', 'e', 'i', 'o', 'u'])
    count = 1
    for s in split:
        tmp = []
        if s[0].lower() in vowels:
            tmp.append(s)
        else:
            tmp.append(s[1:] + s[0])
        tmp.append("ma")
        tmp.append("a" * count)
        count += 1
        res.append("".join(tmp))
    return " ".join(res)
